### ISSUE TITLE
[Предложение] Совместимость с хоткеями Ctrl+Alt+ в расширенных раскладках

### ISSUE THREAD

**Issue by VladWinner** [Aegel5/SimpleSwitcher#54]
### Проблема
Это давняя проблема, но не утилиты, а самой Windows и софта на этой ОС: Ctrl+Alt+[буквы] вместо действий вводят символы из раскладки с альтернативным рядом. Не каждое приложение умеет корректно различать и перехватывать нажатия клавиатуры из-за особенностей работы раскладок в Windows, в Microsoft [ещё в 2004 году](https://devblogs.microsoft.com/oldnewthing/20040329-00/?p=40003) просили разработчиков приложений ничего не назначать на Ctrl+Alt.
### Подробности
C правым Alt можно вставлять спецсимволы, это затрагивает европейские, японские и типографские / расширенные раскладки. При этом Windows интерпретирует Ctrl+Alt как правый Alt, из-за чего вместо нужной комбинации будет ошибочно выставляться какой-нибудь символ в соответствии с выбранной раскладкой. Какой-то системной опции в Windows для отключения интерпретации Ctrl+Alt в качестве замены правого Alt не предусмотрено.
### Пример неправильного поведения
У меня вместо действия на комбинацию Ctrl+Alt+X на типографских раскладках Бирмана выставляется «×».
### Идея
В момент зажатия Ctrl+Alt программа временно подменяет раскладку клавиатуры на ту, где левый и правый Alt равнозначны, например `67699721` - English (US). Раз в базовых раскладках AltGr / правый Alt не признается (нет альтернативного ряда символов), то и комбинации клавиш работают как положено в программах, где додумались их повесить на Ctrl+Alt. После отпускания клавиш возвращается последняя используемая раскладка.

**Comment by Aegel5** [comment]
Эти раскладки: https://ilyabirman.ru/typography-layout/
Например, в последнем Office Excel хоткей Ctrl+Alt+V вставляет символ «↓» вместо открытия окна «Специальной вставки»

**Comment by Aegel5** [comment]
Актуальна еще проблема?

**Comment by VladWinner** [comment]
Да, ещё актуальна!

**Comment by Aegel5** [comment]
Добавил в версию 5.293. Единственное, у меня почему то заработало только с опцией Alternative mode layout change. Соответственно нужно настроить win hotkey для всех раскладок (В самой Windows в том числе)

**Comment by Aegel5** [comment]
1. У меня все работает
2. Галка включить на скриншоте не стоит
3. win hotkey должны быть такими же как выставлены в windows

**Comment by VladWinner** [comment]
![Image](https://github.com/user-attachments/assets/f1eed595-b8f0-4885-a8e7-5d9207047c9d)
![Image](https://github.com/user-attachments/assets/0a2cbc7d-e754-4752-8eec-63e46d87dc8c)
Переключение раскладок работает, но LCtrl+Alt+V снова вызывает печать символа ↓ как при AltGr+V

**Comment by Aegel5** [comment]
Да, нужно настроить, в Win и в SS раскладки отдельно для каждого языка. 
Пытался через это переключать 
https://learn.microsoft.com/ru-ru/windows/win32/api/msctf/nf-msctf-itfinputprocessorprofilemgr-activateprofile
но что-то не особо получилось.

**Comment by Aegel5** [comment]
Еще у вас на скриншоте английская тоже типографская? Разве там нет расширенных значений? Тогда этот хак не будет работать...

**Comment by VladWinner** [comment]
Не всматривался на код раскладки, действительно это моя расширенная (кастомная). 

Тут закономерно, что у пользователя могут быть обе раскладки расширенные, а переключение для хоткеев должно быть всегда на стандартную, то бишь ~ 67699721 — English (US). 
Вообще незачем давать пользователю выбирать из своих кастомных, если корректно переключение сработает только на базовую английскую, которую можно вызывать независимо от того, какие сейчас выставлены раскладки.

Для примера: временное переключение на нужную раскладку при Ctrl+Alt в Mahou
https://gitea.com/BladeMight/Mahou/search?q=CTRL_ALT_TemporaryLayout
https://gitea.com/BladeMight/Mahou/search?q=LCTRLLALTTempLayout

**Comment by Aegel5** [comment]
> которую можно вызывать независимо от того, какие сейчас выставлены раскладки.
В Mahou это работает? Не знал, что можно вызывать раскладку, если она не выставлена.

**Comment by VladWinner** [comment]
Да, в Mahou работает, и при нажатии LCtrl+Alt+V происходит вставка без форматирования в приложениях по работе с текстом:
![Image](https://github.com/user-attachments/assets/2f79d45e-d09d-440d-b608-8bd38d9af30e)

P. S. Mahou из-за перегруженности разными возможностями долго запускается и больше потребляет CPU в фоне при том же минимальном наборе включенных функций, поэтому лично я сейчас пользуюсь SimpleSwitcher.

**Comment by VladWinner** [comment]
> Не всматривался на код раскладки

![Image](https://github.com/user-attachments/assets/44de947a-4d5a-4b28-9146-28e91f6a0dea)
А отсюда никак нельзя извлекать название раскладки?

**Comment by Aegel5** [comment]
> А отсюда никак нельзя извлекать название раскладки?
Можно через winapi находить - это более правильно, реестр может изменяться

**Comment by Aegel5** [comment]
Сбилдил еще версию. В общем, сделал похожим образом как в mahao, вроде действительно можно грузить раскладку даже если она не выставлена. Альтернативный режим теперь не требуется для работы это опции.

**Comment by VladWinner** [comment]
Попробовал 5.302. Пара моментов:
1) Не думаю, что понятно, что означает 4090409 в названии функции, может так и написать, что речь про English (US) раскладку?
2) Переключение на раскладку есть, но её стандартное действие после комбинации клавиш не срабатывает. Такое ощущение, что подмена раскладки происходит в момент удержания трех клавиш, а не двух (Ctrl+Alt), из-за действие не выполняется. Визуально я вижу "ENG US" в области уведомлений только после нажатия дополнительной кнопки.

**Comment by Aegel5** [comment]
Странно, до этого у меня все работало, а теперь стал такой же эффект)
Я так понял раскладка успевает выгрузиться, прежде чем сработает хот-кей - добавил еще один Sleep, сбилдил версию.

**Comment by VladWinner** [comment]
5.306 - то же самое, не срабатывает. Переключение по-прежнему происходит только после трех кнопок, а не первых двух. 
При просмотре с KeyboardTester видно, что при нажатии Ctrl+Alt+V печатается ещё и клавиша Caps (хм?):

https://github.com/user-attachments/assets/701b5dab-0767-4216-ad7a-84b0190283b2

**Comment by Aegel5** [comment]
У меня стабильно отрабатывает. Не знаю, может на win11 поведение другое. Повторите тогда с включенным debug log может так будет понятнее в чем дело.

**Comment by Aegel5** [comment]
Также советую проверить будет ли работать Ctrl + Alt + V если вручную добавить En-us раскладку в систему.

**Comment by VladWinner** [comment]
На v5.319 всё заработало:
[SimpleSwitcher.exe.log](https://github.com/user-attachments/files/19201813/SimpleSwitcher.exe.log)

Только я бы что-нибудь поделал с залипанием, а то удержание Ctrl + Alt + V создаст на несколько секунд подряд «зависание» в раскладке En(US), возможно теперь такое большое ожидание использовать не требуется

**Comment by VladWinner** [comment]
> Ну может это только на моей ОС только такие приколы.

Нет, у меня та же история)
